src/main/java/edu/harvard/iq/dataverse/PermissionServiceBean.java

```diff
@@ -50,347 +53,392 @@
      */
-    public class StaticPermissionQuery {
-
-        final DvObject subject;
-        final RoleAssignee user;
-
-        private StaticPermissionQuery(RoleAssignee user, DvObject subject) {
-            this.subject = subject;
-            this.user = user;
+    public RequestPermissionQuery on(DvObject d) {
+        if (d == null) {
+            throw new IllegalArgumentException("Cannot query permissions on a null DvObject");
         }
-
-        public StaticPermissionQuery user(RoleAssignee anotherUser) {
-            return new StaticPermissionQuery(anotherUser, subject);
+        if (d.getId() == null) {
+            throw new IllegalArgumentException("Cannot query permissions on a DvObject with a null id.");
         }
+        return new RequestPermissionQuery(d, null);
+    }
 
-        /**
-         * "Fast and loose" query mechanism, allowing to pass the command class
-         * name, does not take request-level permissions into account. Command is assumed to live in
-         * {@code edu.harvard.iq.dataverse.engine.command.impl.}
-         *
-         * @deprecated Use DynamicPermissionQuery instead
-         * @param commandName
-         * @return {@code true} iff the user has the permissions required by the
-         * command on the object.
-         * @throws ClassNotFoundException
-         */
-        @Deprecated
-        public boolean canIssueCommand(String commandName) throws ClassNotFoundException {
-            return isUserAllowedOn(user,
-                    (Class<? extends Command>) Class.forName("edu.harvard.iq.dataverse.engine.command.impl." + commandName), subject);
+    /**
+     * Start querying permissions for {@link DataverseRequest} {@code req}. 
+     * @param req the {@link DataverseRequest} whose permissions will be queried
+     * @return A permission query
+     */
+    public RequestPermissionQuery request(DataverseRequest req) {
+        return new RequestPermissionQuery(null, req);
+    }
+    
+    /**
+     * Checks whether a command has enough permissions to run.
+     * 
+     * Note that the command may fail due to other reasons, such as internal 
+     * errors or making no sense (see {@link IllegalCommandException}).
+     * 
+     * @param aCommand
+     * @return {@code true} iff submitting the command will not cause a {@link PermissionException} to be thrown.
+     */
+    public boolean isPermitted( Command aCommand ) {
+        return isPermitted(aCommand, null);
+    }
+    
+    /**
+     * Internal call; checks permissions, allows a callback for fine-grained
+     * response if case there are not enough permissions.
+     * 
+     * @param aCommand The command to be tested/
+     * @param ipc Callback, called if there are insufficient permissions for the passed command.
+     * @return {@code true} iff the command will not be blocked due to permission issues.
+     */
+    boolean isPermitted(Command aCommand, InsufficientPermissionCallback ipc) {
```

Comment:
Does anything use this anymore? If not, we should remove.

Reply:
Yes, used twice (engine and internally in the bean).